File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int hold = -prices[0], free = 0;
        for(int i=1;i<n; i++)
        {
            hold = max(hold, free - prices[i]);
            free = max(free, prices[i] + hold);
        }
        return free;
    }
};

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n <= 1: return 0
        res = 0
        for i in range(1, n):
            res += max(0, prices[i]-prices[i-1])
        return res